### DISCUSSION TITLE
RuntimeError: not a dict (null)

### DISCUSSION THREAD

**Discussion by doppelmarker** [pymupdf/PyMuPDF#1764] [Looking for help] [answered]
Whenever I try to iterate over the pages of 2 pdf-documents attached and achieve the page with form I get this error:

![image](https://user-images.githubusercontent.com/47917211/174572181-b3722570-dce2-4a0b-8e7f-a4419c962b05.png)

The common thing between them is that both contain Signature fields, maybe the problem is here? Thanks.

[failed_pdf1.pdf](https://github.com/pymupdf/PyMuPDF/files/8939227/failed_pdf1.pdf)
[failed_pdf2.pdf](https://github.com/pymupdf/PyMuPDF/files/8939389/failed_pdf2.pdf)

**Comment by JorjMcKie** [comment, marked as answer]
As was to be expected: your files contain errors:
```python
for page in doc:
    print("Page", page.number)

    
Page 0
Traceback (most recent call last):
  File "<pyshell#31>", line 1, in <module>
    for page in doc:
    <...>
    val = _fitz.Document_load_page(self, page_id)
RuntimeError: not a dict (null)
# see what MuPDF has to say:
print(fitz.TOOLS.mupdf_warnings())
invalid indirect reference (0 0 R)
... repeated 2 times...
not a dict (null)
cannot create appearance stream
invalid indirect reference (0 0 R)
... repeated 2 times...
not a dict (null)
```
If you look at the file directly with some text editor, you will see something like this:
![grafik](https://user-images.githubusercontent.com/8290722/174608112-eaa67777-cc13-41db-9b53-ed96da921df4.png)

By inspecting all xref sources with PyMuPDF, you could also confirm this:
```python
for xref in range(1, doc.xref_length()):
    if " 0 0 R" in doc.xref_object(xref):
        print(f"Error '0 0 R' in object definition {xref}!")

        
Error '0 0 R' in object definition 27!
Error '0 0 R' in object definition 31!
Error '0 0 R' in object definition 32!
Error '0 0 R' in object definition 33!
Error '0 0 R' in object definition 34!
Error '0 0 R' in object definition 35!
Error '0 0 R' in object definition 39!
Error '0 0 R' in object definition 40!
Error '0 0 R' in object definition 41!
Error '0 0 R' in object definition 42!
Error '0 0 R' in object definition 44!
```

You must clean the file before you can use a page which refers to an illegal xref.

**Reply by JorjMcKie** [reply]
"Cleaning" works like so:

* either do a separate CLI step `mutool clean -ggg error.pdf clean.pdf` (use a garbage option that causes MuPDF to inspect all xrefs), then work with `clean.pdf`,
* or dynamically do the same in your script:
```python
doc=fitz.open("error.pdf")
pdfbytes=doc.tobytes(garbage=3)
doc.close()
doc=fitz.open("pdf", pdfbytes)
```

**Reply by doppelmarker** [reply]
Thanks a lot, this fixed the issue! Actually, I was using 1.19.6 version whereas 1.20.0 was released 5 days ago. Updating to a new version resolves the problem as well.

**Reply by PasaOpasen** [reply]
@JorjMcKie is there some fast way to check if documents needs cleaning?

**Reply by JorjMcKie** [reply]
> @JorjMcKie is there some fast way to check if documents needs cleaning?

Either an error is detectable at open, then this is fast *per se*, or it requires actually trying to read/interpret an object - then this is slow.
Even reading a PDF's pages from front to end is no guarantee to detect every error.
